### PR TITLE
Fix contextual discrimination for omitted members

### LINKED ISSUE #41759: Omitted property does not narrow discriminated union contextual type
**TypeScript Version:**  4.2.1

**Search Terms:** discriminated union optional property

**Code**

```ts repro
type DiscriminatorTrue = {
  disc: true;
  cb: (x: string) => void;
}

type DiscriminatorFalse = {
  disc?: false;
  cb: (x: number) => void;
}

type Props = DiscriminatorTrue | DiscriminatorFalse;

declare function f(options: DiscriminatorTrue | DiscriminatorFalse): any;

f({
  disc: true,
  cb: s => parseInt(s) // Inference works 👍 
});

f({
  disc: false,
  cb: n => n.toFixed() // Inference works 👍 
});

f({
  cb: n => n.toFixed() // Implicit any error 👎 
});

f({
  cb: (n: string) => parseInt(n) // But errors correctly with incorrect type annotation 👍 
});
```

**Expected behavior:**
No implicit any error on the third call site

**Actual behavior:**
Implicit any error on the callback parameter at the third call site

[**Playground Link**](https://www.typescriptlang.org/play?ts=4.1.2#code/C4TwDgpgBAIglgZwMYCc4Fs4DsCGwD2KAKigK7QC8UA3gFBRQAmiSAXFMGRANz1RIAjdgAoAHuwSdsAcwCUUCgD4oAN3xxGvAL61aoSLBZpMuAigBiOADYJKNPs2QB+dgDNrt3g0EjxULKToAhAo8kqq6pq0Onrg0AAKKPhgCAqGyMbYeIQk5FAAPumoGFlmljY8uowQSFY4KNCupFhIwHD4WFCuwsltHQjs8BklpjlcBUWZoxYeELLsOFggvLTddAyObBxcADR8PlCp4WD1tgCSWMDCCPIA9LdQF64hEC3QAO6EANYI0bIrawcLDcsz23iE-gUyiwADoCOY4KIIIxhHcHk8Xm8oJ8UD8-gDhOt+BDOuFYfDEcjUVB7o90GArHAkHBgFBFiB8bpAeCRFgJFIsHIoVATihzpdhFg0VAAEKkVkhJJi-iEBqtKwgbEsgAWUGwSFVNVZ+mgiyw+GAeHaWE5QA)

**Related Issues:** https://github.com/microsoft/TypeScript/issues/31404#issuecomment-596841068 is this issue, but the OP there is different.

**Notes:**
This pattern sometimes appears in React component props, where convention is to make boolean properties optional and only pass them as `true` (usually with the shorthand `<MyComponent boolProp />`). I actually suggested using discriminated union props for React components in [an old blog post](https://blog.andrewbran.ch/expressive-react-component-apis-with-discriminated-unions/), and noted this issue in a [footnote](https://blog.andrewbran.ch/expressive-react-component-apis-with-discriminated-unions/#fn-1), calling it a possible bug, but didn’t file it at the time because I had low confidence it wasn’t a duplicate or design limitation. It was later mentioned in https://github.com/microsoft/TypeScript/issues/31404#issuecomment-596841068, but was probably ignored because it was assumed to be an instance of the OP’s issue, which was determined to be a design limitation. I dove into this again because [someone tweeted at me](https://twitter.com/ChrisDascoli5/status/1333123689139855362) asking about that footnote after reading the post.

### PR DESCRIPTION
**Pull request by erikbrinkman**
In my original PR #43633, choices for optional properties that could be discriminatory were chosen from the union type, which in turn meant that the possible omitted properties to check for discrimination was taken from the intersection of properties across all types. This is the incorrect way to check, when it's actually the union of all properties that should be checked for discrimination. See the additions to the tests for details.

Two stylistic notes about this PR. In principle the code should be going from intersection to union, but it proved more difficult to do given the current functions I found available. As a result:
1. I couldn't find the best way to get the unique strings of an array like `[...new Set()]` would, so I created a small helper function, but didn't put it in a great place. Given the lack of documentation in checker, I'm not sure where to put it, or if there are better utilities I can use to achieve the same effect.
2. This function was relatively hard to parse before. Arguably the first PR didn't help, but it kept it at roughly the same complexity. This PR makes it much worse. Ideally I'd like to strip out some common functionality, or decompose the nested loops, but from looking through checker that seems less the preferred style. Basically, this works, but I don't love it, and feedback is welcome.

Given your initial suggestions and the review of the past PR, @weswigham seems poised to review this the quickest.

Fixes #41759

### CODE REVIEW

**Comment by erikbrinkman** [comment]
I'm not sure how to interact with typescript bot, but the previous PR seems to have introduced a stack overflow, which should be addressed at least in tandem with this fix.

**Comment by erikbrinkman** [comment]
Hey @weswigham , the one user who reported a stack overflow hasn't followed up to debug, so I'm hoping it's a non issue. I don't love the style of this code, primarily because it's not clear how best to utilize sets in `checker.ts`, or when it's reasonable to factor out functions, or where to put utilities like `uniqueStrings`. Any general comments you have on this, or the meat of the commit would be much appreciated.

**Comment by DanielRosenwasser** [comment]
This PR seems to have regressed our check times on material UI drastically.

image
